### PR TITLE
Fix DDP no_sync when find_unused_parameters is True

### LINKED ISSUE #69031: DDP only syncs parameters used in most recent pass when `find_unused_parameters` is True.
## 🐛 Bug

I'm working with a training loop with gradient accumulation, where several forwards/backwards passes run for each call to the optimizer step. To reduce communication overhead, DDP syncs are disabled for all but the final gradient accumulation step before the optimizer step. I'm also using an optimized model which randomly disables various parts of itself on each pass. Critically, the parts of the model that are disabled can vary between gradient accumulation steps. The problem I'm observing is that the DDP sync ran on that final grad accum step appears to only be syncing the parameters used on that step. Parameters unused on the final step but used on previous steps are never synced.

## To Reproduce

```python
import contextlib
import os

import torch
import torch.distributed
import torch.multiprocessing
import torch.nn as nn

GRADIENT_ACCUMULATION_STEPS = 2


class SimpleConditionalModel(nn.Module):
    # uses nn1 layer on the first pass; nn2 layer on the second pass

    def __init__(self):
        super().__init__()

        self.nn1 = nn.Linear(1, 1)
        self.nn2 = nn.Linear(1, 1)

        self.loss = nn.MSELoss()

        self.state = 0

    def forward(self, input):
        if self.state == 0:
            self.state = 1
            return self.nn1(input)
        else:
            self.state = 0
            return self.nn2(input)


def test_model(rank):
    os.environ["RANK"] = str(rank)
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "29000"

    torch.distributed.init_process_group(backend='gloo', world_size=2)
    model = SimpleConditionalModel()
    ddp = torch.nn.parallel.DistributedDataParallel(model, find_unused_parameters=True)

    for microbatch_idx in range(GRADIENT_ACCUMULATION_STEPS):
        context = contextlib.nullcontext
        if microbatch_idx < GRADIENT_ACCUMULATION_STEPS - 1:
            context = ddp.no_sync

        with context():
            input = torch.rand((1, ))
            output = ddp.forward(input)
            target = torch.rand((1, ))

            loss = model.loss(output, target)
            loss.backward()

    print(f'Rank {rank} gradients: {[p.grad for p in model.parameters()]}')


if __name__ == '__main__':
    torch.multiprocessing.spawn(test_model, nprocs=2)
```

Output:
```
Rank 0 gradients: [tensor([[-1.1942]]), tensor([-1.4291]), tensor([[-0.4950]]), tensor([-3.3037])]
Rank 1 gradients: [tensor([[-0.3461]]), tensor([-0.4722]), tensor([[-0.4950]]), tensor([-3.3037])]
```
The first two gradients, corresponding to the `nn1` layer, are unsynced.

## Expected behavior

DDP should be able to track used/unused parameters across multiple forwards/backwards passes.

## Environment
 - PyTorch Version: 1.9.0
 - OS: Linux
 - How you installed PyTorch: `pip`
 - Python version: 3.8.0

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by cbalioglu**
Fixes #69031, #42793

This PR fixes the bug introduced in #54981 where parameters used within a `no_sync` scope are not respected when `find_unused_parameters` is set to `True`. The `local_used_map_` and `numGradHooksTriggeredMap_` variables should be updated regardless of the `no_sync` state.

Tested and verified with fairseq2 and wav2vec2 ASR finetuning recipe. All gradients are correctly synced across workers as expected after applying this fix.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Looks good, but can we get a unittest?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -7781,6 +7781,81 @@ def forward(_self, input, expected_type):  # noqa: B902
             inp = TestNamedTupleInput_1(a, b)
             model(inp, type(inp))
 
+        @require_backend_is_available({"gloo"})
+        def test_grads_same_across_ranks_with_no_sync(self):
+            group, group_id, rank = self._init_global_test()
+            world_size = dist.get_world_size()
+            if world_size < 2:
+                self.skipTest("This test requires at least two ranks.")
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:7789]: nit: you can grep around this file to find decorators that skip with these sort of conditions.

**Review by rohan-varma — APPROVED** [review]
Thank you!

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (kauterry) [reply torch/testing/_internal/distributed/distributed_test.py:7789]: I did check, this exact condition is not met. There is a decorator with <= 2 `skip_if_small_worldsize`, but this test should be run with 2 ranks as well.
